### PR TITLE
Add `dropout_edge` and deprecate `dropout_adj`

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
This is a follow-up PR for https://github.com/pyg-team/pytorch_geometric/issues/5452. This PR added `dropout_edge` and deprecated `dropout_adj`. 
TODO: Make `edge_mask` align with `edge_index` when `force_undirected=True`

I also fixed the doc-string in `dropout_node`, as it returned the `edge_mask` and `node_mask` indicate where edges and nodes are retained rather than dropped/masked. Sorry for the mess!

### CODE REVIEW

**Comment by rusty1s** [comment]
Cool! Can we also remove any occurrence of `dropout_adj` with `dropout_edge` in PyG?

**Comment by EdisonLeeeee** [comment]
I am struggling to align `edge_mask` and `edge_index` when `force_undirected=True`. It seems hard to get the `edge_mask` after the `edge_index` has been made symmetric. @rusty1s Could you please take a look and give me some help?

File: torch_geometric/utils/dropout.py [hunk 1]
```diff
@@ -135,3 +137,62 @@ def dropout_node(edge_index: Tensor, p: float = 0.5,
...
+        return edge_index, edge_mask
+
+    row, col = edge_index
+
+    edge_mask = torch.rand(row.size(0), device=edge_index.device) >= p
+
+    if force_undirected:
+        edge_mask[row > col] = False
+
+    row, col, _ = filter_adj(row, col, None, edge_mask)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:189]: Can we remove the `filter_adj` function and simply do
```python
edge_index = edge_index[:, edge_mask]
```
This avoids the extra `stack` later.

File: torch_geometric/utils/dropout.py [hunk 2]
```diff
@@ -135,3 +137,62 @@ def dropout_node(edge_index: Tensor, p: float = 0.5,
...
+
+    if force_undirected:
+        edge_index = torch.stack(
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:192]: Here we can then do
```python
edge_index = torch.cat([edge_index, edge_index.flip([0])], dim=1)
```

**Comment by rusty1s** [comment]
Can we also replace the occurrence of `dropout_adj` in the `README.md` with `dropout_edge`?

File: torch_geometric/utils/dropout.py [hunk 3]
```diff
@@ -135,3 +137,62 @@ def dropout_node(edge_index: Tensor, p: float = 0.5,
...
+
+    if not training or p == 0.0:
+        edge_mask = edge_index.new_zeros(edge_index.size(1), dtype=torch.bool)
+        return edge_index, edge_mask
+
+    row, col = edge_index
+
+    edge_mask = torch.rand(row.size(0), device=edge_index.device) >= p
+
+    if force_undirected:
```
[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:192]: This seems cannot guarantee that the output `edge_index` could be selected by masking input with `edge_mask`, i.e.,:
```python
out_edge, edge_mask = dropout_edge(in_edge)
# out_edge != in_edge[edge_mask]
```

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/utils/dropout.py:192]: Oh, you are right. We need to return indices here.

[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:192]: Still cannot find a better solution for this. How about dropping this argument `force_undirected`? One can use `to_undirected` after that if an undirected graph is needed. Or he can use `dropout_adj` instead.

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/utils/dropout.py:192]: Wouldn't it be okay to return indices here?

[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:192]: How can we find the indices of those reversed edges? We can easily find that of `(row, col)` while it is challenging to find that of `(col, row)`.

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/utils/dropout.py:192]: We just need the indices from the lower triangular matrix (this is what the current implementation of `dropout_adj` does).

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/utils/dropout.py:192]: Yes, I fully agree, but it's better than not returning anything for `force_undirected=True`. We can still return masks for the other cases. Not ideally, but better than anything else?

[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/utils/dropout.py:192]: Makes sense. Let me have a try again :)

File: README.md [hunk 4]
```diff
@@ -307,7 +307,8 @@ New Benchmarks and Strong Simple Methods](https://arxiv.org/abs/2110.14446) (Neu
 PyG comes with a rich set of neural network operators that are commonly used in many GNN models.
 They follow an extensible design: It is easy to apply these operators and graph utilities to existing GNN layers and models to further enhance model performance.
 
-* **[DropEdge](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.dropout_adj)** from Rong *et al.*: [DropEdge: Towards Deep Graph Convolutional Networks on Node Classification](https://openreview.net/forum?id=Hkx1qkrKPr) (ICLR 2020)
+* **[DropEdge](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.dropout_edge)** from Rong *et al.*: [DropEdge: Towards Deep Graph Convolutional Networks on Node Classification](https://openreview.net/forum?id=Hkx1qkrKPr) (ICLR 2020)
+* **[DropNode](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.dropout_node)** from You *et al. * [Graph Contrastive Learning with Augmentations](https://arxiv.org/abs/2010.13902) (NeurIPS 2020)
```
Comment (rusty1s) [inline README.md:311]:
```suggestion
* **[DropNode](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.dropout_node)** from You *et al.* [Graph Contrastive Learning with Augmentations](https://arxiv.org/abs/2010.13902) (NeurIPS 2020)
```

File: torch_geometric/utils/dropout.py [hunk 5]
```diff
@@ -135,3 +143,68 @@ def dropout_node(edge_index: Tensor, p: float = 0.5,
...
+                [1, 2, 1, 3]])
+        >>> edge_mask # masks indicating which edges are retained
+        tensor([ True, False,  True,  True,  True, False])
+
+        >>> edge_index, edge_mask = dropout_edge(edge_index,
+        ...                                      force_undirected=True)
+        >>> edge_index
+        tensor([[0, 1, 2, 1, 2, 3],
+                [1, 2, 3, 0, 1, 2]])
+        >>> edge_mask # indices indicating which edges are retained
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:186]:
```suggestion
        >>> edge_index, edge_id = dropout_edge(edge_index,
        ...                                      force_undirected=True)
        >>> edge_index
        tensor([[0, 1, 2, 1, 2, 3],
                [1, 2, 3, 0, 1, 2]])
        >>> edge_id # indices indicating which edges are retained
```

**Review by rusty1s — APPROVED** [review]
Beautiful, thank you!

File: test/utils/test_dropout.py [hunk 6]
```diff
@@ -41,3 +41,21 @@ def test_dropout_node():
...
     assert out[1].tolist() == [False, False, False, False, True, True]
     assert out[2].tolist() == [True, False, True, True]
+
+
+def test_dropout_edge():
+    edge_index = torch.tensor([[0, 1, 1, 2, 2, 3], [1, 0, 2, 1, 3, 2]])
+
+    out = dropout_edge(edge_index, training=False)
+    assert edge_index.tolist() == out[0].tolist()
+    assert out[1].tolist() == [False, False, False, False, False, False]
```
Comment (rusty1s) [inline test/utils/test_dropout.py:51]:
```suggestion
    assert out[1].tolist() == [True, True, True, True, True, True]
```

File: torch_geometric/utils/dropout.py [hunk 7]
```diff
@@ -135,3 +143,68 @@ def dropout_node(edge_index: Tensor, p: float = 0.5,
...
+                [1, 2, 3, 0, 1, 2]])
+        >>> edge_id # indices indicating which edges are retained
+        >>> tensor([0, 2, 4, 0, 2, 4])
+    """
+    if p < 0. or p > 1.:
+        raise ValueError(f'Dropout probability has to be between 0 and 1 '
+                         f'(got {p}')
+
+    if not training or p == 0.0:
+        edge_mask = edge_index.new_zeros(edge_index.size(1), dtype=torch.bool)
```
Comment (rusty1s) [inline torch_geometric/utils/dropout.py:194]:
```suggestion
        edge_mask = edge_index.new_ones(edge_index.size(1), dtype=torch.bool)
```
